vic/extensions/rout_rvic/src/rout_run.c

```diff
@@ -0,0 +1,142 @@
+    for (i = 0; i < local_domain.ncells_active; i++) {
+        dvar_local_total_runoff[i] = out_data[i][OUT_RUNOFF][0] +
+                                     out_data[i][OUT_BASEFLOW][0];
+    }
+
+    gather_put_var_double(dvar_global_total_runoff, dvar_local_total_runoff);
+
+    free(dvar_local_total_runoff);
+
+    // Do the routing on the master node
+    if (mpi_rank == VIC_MPI_ROOT) {
+        debug("RVIC");
+
+        dvar_global = malloc(global_domain.ncells_total * sizeof(*dvar_global));
+        check_alloc_status(dvar_global, "Memory allocation error.");
+
+        // Zero out current ring
+        // in python: (from variables.py) self.ring[tracer][0, :] = 0.
+        for (i_outlet = 0; i_outlet < rout.rout_param.n_outlets; i_outlet++) {
+            rout.ring[i_outlet] = 0.0;
+        }
+
+        // Equivalent to Fortran 90 cshift function, in python: (from variables.py)
+        // self.ring[tracer] = np.roll(self.ring[tracer], -1, axis=0)
+        cshift(rout.ring, rout.rout_param.full_time_length,
+               rout.rout_param.n_outlets, 0,
+               1);
+
+        /*Loop through all sources*/
+        for (i_source = 0; i_source < rout.rout_param.n_sources; i_source++) {
+            i_outlet = rout.rout_param.source2outlet_ind[i_source];
+            offset = rout.rout_param.source_time_offset[i_source];
+
+            /* Do the convolution */
+            // iTimestep is the position in the unit hydrograph
+            // jTimestep is the position in the ring
+            for (i_timestep = 0; i_timestep < rout.rout_param.n_timesteps;
+                 i_timestep++) {
+                j_timestep = i_timestep + offset;
+
+                // index locations
+                i_ring = (j_timestep * rout.rout_param.n_outlets) + i_outlet;
+                i_uh = (i_timestep * rout.rout_param.n_sources) + i_source;
+
+                rout.ring[i_ring] += rout.rout_param.unit_hydrograph[i_uh] *
+                                     dvar_global_total_runoff[rout.rout_param.
+                                                              source_VIC_index[
+                                                                  i_source]];
+            }
+        }
+
+        // Write to dvar_global prior to scattering over local domains...
+        for (i_outlet = 0; i_outlet < rout.rout_param.n_outlets; i_outlet++) {
+            dvar_global[rout.rout_param.outlet_VIC_index[i_outlet]] =
+                rout.ring[i_outlet] *
+                global_domain.locations[rout.rout_param.outlet_VIC_index[
+                                            i_outlet]].area /
+                (MM_PER_M * global_param.dt);
+        }
+
+        // cleanup
+        free(dvar_global);
+    }
+
+    // Scatter dvar_global to local domains (dvar_local)...
+    dvar_local = malloc(local_domain.ncells_active * sizeof(*dvar_local));
+    check_alloc_status(dvar_local, "Memory allocation error.");
+
+    get_scatter_var_double(dvar_global, dvar_local);
```

Comment:
I'm having a bit of trouble following some of this memory logic but I think `dvar_global` is possibly freed prematurely.  Perhaps using more descriptive names would be good, and please check that this memory is allocated/freed correctly.